File: Solutions/147_Insertion_Sort_List/147_Insertion_Sort_List.go
func insertionSortList(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }
    result := &ListNode{Next: head}
    cur := head.Next
    head.Next = nil
    
    for cur != nil {
        pre := result
        target := result.Next
        for target != nil && cur.Val > target.Val {
            target = target.Next
            pre = pre.Next
        }
        temp := cur
        cur = cur.Next
        temp.Next = target
        pre.Next = temp
    }
    
    return result.Next
}

//Fourth fastest
func insertionSortList(head *ListNode) *ListNode {
    if head == nil {
        return head
    }
    var arr []int
    
    for head != nil {
        arr = append(arr, head.Val)
        head = head.Next
    }
    for i := 1; i < len(arr); i++ {
        for j := i; j > 0; j-- {
            if arr[j] > arr[j-1] {
                break
            }
            arr[j-1], arr[j] = arr[j], arr[j-1]
        }
    }
    
    head2 := &ListNode{arr[0], nil}
    prev := head2
    
    for i := 1; i < len(arr); i++ {
        next := &ListNode{arr[i], nil}
        prev.Next = next
        prev = next
    }
    
    return head2
}

//Third fastest
func insertionSortList(head *ListNode) *ListNode {
    
    if(head == nil) {
        return nil
    }
    
    curr:=head.Next
    prev:= head
    
    for (curr!=nil) {
        
        if (curr.Val < prev.Val) {
            prev.Next = curr.Next
            
            var prev2 *ListNode
            curr2 := head
            
            for (curr2!=nil) {
                if (curr2.Val > curr.Val) {
                    curr.Next = curr2
                    if(prev2 != nil) {
                        prev2.Next = curr
                    }else {
                        head = curr
                    }
                    break
                }else {
                    prev2 =curr2
                    curr2=curr2.Next
                }
                
                
            }
            curr = prev.Next
            
        } else {
            prev = curr
            curr = curr.Next
        }
        
    }
    return head
}

//Second fastest
func insertionSortList(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }
    
    dummy := &ListNode{Next:head}
    cur := head
    for cur != nil && cur.Next != nil {
        next := cur.Next        
        if cur.Val > next.Val {            
            iter := dummy
            for iter.Next != nil && iter.Next.Val < next.Val {
                iter = iter.Next
            }
            cur.Next = next.Next
            next.Next = iter.Next
            iter.Next = next
        } else {
            cur = next
        }
    }
    return dummy.Next
}

//Fastest
func insertionSortList(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }
    
    dummy := &ListNode{Next:head}
    cur := head
    for cur != nil && cur.Next != nil {
        next := cur.Next        
        if cur.Val > next.Val {            
            iter := dummy
            for iter.Next != nil && iter.Next.Val < next.Val {
                iter = iter.Next
            }
            cur.Next = next.Next
            next.Next = iter.Next
            iter.Next = next
        } else {
            cur = next
        }
    }
    return dummy.Next
}

File: Solutions/147_Insertion_Sort_List/147_Insertion_Sort_List.py
class Solution:
    def insertionSortList(self, head):
        dummy = ListNode(-1)
        curr = head
        while curr:
            curr_next = curr.next
            prv, nxt = dummy, dummy.next
            while nxt:
                if nxt.val > curr.val: break
                prv = nxt
                nxt = nxt.next
                
            curr.next = nxt
            prv.next = curr
            curr = curr_next
        
        return dummy.next

#Most common
class Solution:
    def insertionSortList(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        
        dummy = ListNode(0, head)
        tail = head
        
        while tail.next:
            insert_node = tail.next
            
            if tail.val <= insert_node.val:
                tail = tail.next
            else:
                tail.next = insert_node.next

                temp = dummy
                while temp.next.val < insert_node.val:
                    temp = temp.next
            
                insert_node.next = temp.next
                temp.next = insert_node
            
                # tail = tail.next
        return dummy.next

#Third fastest
class Solution:
    def insertionSortList(self, head: ListNode) -> ListNode:
        if not head:
            return None
        node_list = []
        while head:
            node_list.append(ListNode(head.val))
            head = head.next
        
        def output_val(node):
            return node.val
        node_list.sort(key = output_val)
        
        for i, item in enumerate(node_list):
            item.next = node_list[i+1] if i<len(node_list)-1 else None
        return node_list[0]

#Second fastest
class Solution:
    def insertionSortList(self, head: ListNode) -> ListNode:
        nodeDict=[]
        res = pre = ListNode(0)
        while head:
            nodeDict.append([head.val,head])
            head = head.next
        sortedDict = sorted(nodeDict,key=(lambda x:x[0]))
        for key, value in sortedDict:
            pre.next = value
            pre = pre.next
        pre.next = None
        return res.next
        

#Fastest
class Solution:
    def insertionSortList(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        
        li = []
        
        work = head
        while work:
            li.append(work.val)
            work = work.next
            
        li.sort()
        
        work = head
        
        for i in range(len(li)):
            work.val = li[i]
            work = work.next
            
        return head